text/projects/Krzysztof_Michalik/Wee_LCP.tex

```diff
@@ -0,0 +1,179 @@
+
+Pozostaje pokazać jak wyliczamy $LCP[i]$ korzystając z $S$, $S_A$ i $T$. Niech $SUF[j] = i$. Wzory na to są następujące:
+
+%Pokażemy, że $LCP[i] = rank_0(S, select_1(S, SUF[i])) - SUF[i] \leftrightarrow LCP[i] + SUF[i] = %rank_0(S, select_1(S, SUF[i])) \leftrightarrow LCP[i] + SUF^{-1}[j] = rank_0(S, select_1(S, SUF[i]))$
+%W tym celu zaobserwujemy następujące fakty:
+
+\begin{enumerate}
+    \item $LCP[i] = rank_0(S, select_1(S, A[i])) - A[i]$
+    \item $LCP[i] = select_1(S, A[i]) - 2A[i]$
+\end{enumerate}
+
+%This is because the select1-statement gives the position where the encoding for H[A[i]] ends in H,
+%and the rank0-statement counts the sum of the I[j]’s for 1 ≤ j ≤ A[i]. So subtracting the value
+%A[i], which has been “artificially” added to the LCP-array, yields the correct value.
+
+Pierwszy wzór jest poprawny, ponieważ $select_1(S, A[i])$ zwraca pozycję, gdzie kończy się kodowanie $LCP[S_A[i]]$ w $S$. Wtedy $rank_0()$ zlicza sumę $DIFF[j]$, dla $j \le A[i]$. Zatem odjęcie $A[i]$ (które było sztucznie dodane aby zagwarantować monotoniczność ciągu), zwróci rzeczywistą wartość $LCP[i]$.
+
+Drugi wzór jest uproszczeniem pierwszego, ponieważ wiemy ile dokładnie występuje biót '1' w tym podstringu, możemy więc policzyć bity '0' bez odwoływania się do zapytania $rank_0()$.
+
+\section{Kompresja LCP do pamięci $O(\frac{n}{\log{\log{n}}}) = o(n)$}
+
+Zauważmy, że jedyna funkcjonalność wymagana od konstrukcji opisanej w pierwszym wyniku, to odpowiadanie na zapytania postaci $select_1(S, i)$, żeby umieć odpowiedzieć na zapytanie o wartość $LCP[i]$. Oznacza to, że wystarczy zmienić strukturę na zajmującą mniejszą asymptotycznie pamięć, która również będzie potrafić odpowiadać na tego typu zapytania, aby dostać lepszy pamięciowo rezultat.
+
+Autor proponuje trochę 'brutalne' podejście, które sprawia, że pamięć potrzebna zostaje zredukowana do $o(n)$, a sam czas ulega wydłużeniu, ponieważ będziemy musieli za każdym razem wykonać kilka porównań w tekście $T$. Jednak ze względu na czas potrzebny na wyliczenie $S_A[i]$ (wynoszący $A_t$ ze względu na kompresję), jeśli dobierzemy odpowiednie stałe, czas asymptotycznie pozostaje taki jak poprzednio.
+
+Rozważmy podział $S$ z poprzedniej sekcji na takie przedziały, że ostatnim elementem każdego przedziału będzie bit '1' oraz każdy z przedziałów będzie zawierał $\kappa = \lfloor \log{n}^2 \rfloor$ bitów '1' (być może poza ostatnim). Oznacza to, ze utworzymy w ten sposób co najwyżej $\frac{n}{\kappa}$ przedziałów.
+
+Teraz będziemy chcieli zapamiętać 2 wartości dla każdego przedziału:
+\begin{enumerate}
+    \item $N[i] = select_1(S, i\kappa)$.
+    \item Jeśli przedział jest dłuższy niż $\kappa^2$, czyli $\kappa^2 > N[i] - N[i-1]$, to zhadrcodujemy odpowiedzi na wszystkie możliwe zapytania $select_1()$ na tym przedziale w tablicy $P$ i zapiszemy wskaźnik do niej.
```

Comment:
zhadrcodujemy, zhardcodujemy -> please change all these to e.g. "zakodujemy jako stałe"